static/app/views/explore/hooks/useGetTraceItemAttributeKeys.tsx

```diff
@@ -0,0 +1,179 @@
+import type {PageFilters} from 'sentry/types/core';
+import type {Tag, TagCollection} from 'sentry/types/group';
+import {FieldKind} from 'sentry/utils/fields';
+import useApi from 'sentry/utils/useApi';
+import useOrganization from 'sentry/utils/useOrganization';
+import usePageFilters from 'sentry/utils/usePageFilters';
+import type {
+  TraceItemDataset,
+  UseTraceItemAttributeBaseProps,
+} from 'sentry/views/explore/types';
+
+interface UseGetTraceItemAttributeKeysProps extends UseTraceItemAttributeBaseProps {
+  datetime?: PageFilters['datetime'];
+  projectIds?: Array<string | number>;
+}
+
+type TraceItemAttributeKeyOptions = Pick<
+  ReturnType<typeof normalizeDateTimeParams>,
+  'end' | 'start' | 'statsPeriod' | 'utc'
+> & {
+  attributeType: 'string' | 'number';
+  itemType: TraceItemDataset;
+  project?: string[];
+  substringMatch?: string;
+};
+
+function makeTraceItemAttributeKeysQueryOptions({
+  traceItemType,
+  type,
+  datetime,
+  projectIds,
+  search,
+}: {
+  traceItemType: TraceItemDataset;
+  type: 'string' | 'number';
+  datetime?: PageFilters['datetime'];
+  projectIds?: Array<string | number>;
+  search?: string;
+}): TraceItemAttributeKeyOptions {
+  const options: TraceItemAttributeKeyOptions = {
+    itemType: traceItemType,
+    attributeType: type,
+  };
+
+  if (search) {
+    options.substringMatch = search;
+  }
+
+  if (projectIds?.length) {
+    options.project = projectIds.map(String);
+  }
+
+  if (datetime) {
+    const {end, start, statsPeriod, utc} = normalizeDateTimeParams(datetime);
+    if (end) {
+      options.end = end;
+    }
+    if (start) {
+      options.start = start;
+    }
+    if (statsPeriod) {
+      options.statsPeriod = statsPeriod;
+    }
+    if (utc) {
+      options.utc = utc;
+    }
+  }
+
+  // environment left out intentionally as it's not supported
+  return options;
+}
+
+export function useTraceItemAttributeKeysQueryOptions({
+  traceItemType,
+  type,
+  datetime,
+  projectIds,
+  search,
+}: {
+  traceItemType: TraceItemDataset;
+  type: 'string' | 'number';
+  datetime?: PageFilters['datetime'];
+  projectIds?: Array<string | number>;
+  search?: string;
+}): TraceItemAttributeKeyOptions {
+  const {selection} = usePageFilters();
+
+  return useMemo(() => {
+    return makeTraceItemAttributeKeysQueryOptions({
+      traceItemType,
+      type,
+      datetime: datetime ?? selection.datetime,
+      projectIds: projectIds ?? selection.projects,
+      search,
+    });
+  }, [selection, traceItemType, type, datetime, projectIds, search]);
+}
+
+export function useGetTraceItemAttributeKeys({
+  traceItemType,
+  projectIds,
+  datetime,
+  type,
+}: UseGetTraceItemAttributeKeysProps) {
+  const api = useApi();
+  const organization = useOrganization();
+  const {selection} = usePageFilters();
+
+  const getTraceItemAttributeKeys = useCallback(
+    async (queryString: string): Promise<TagCollection> => {
```

Comment:
nit: making default to an empty string might make it a little nicer to use

```suggestion
    async (queryString: string = ''): Promise<TagCollection> => {
```